### PR TITLE
Better annotate interprocess_exception

### PR DESCRIPTION
**Pull request by Dr15Jones**
The IBs with the external generator process are periodically failing with the unhelpful error message
```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	26077_0 process: caught exception 
	boost::interprocess_exception::library_error 15-Dec-2022 03:59:39 CET
	  while serialize lumi
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

That message happens when `boost::interprocess_exception` constructor is passed an error_info that is set to 0 AND the constructor is not passed a `const char*`. I went and modified all the places our code might reach a function that calls `boost::interprocess_exception` and forces it to also pass a `const char*` saying why the exception is being thrown. 

Hopefully this will help track down the problems we see ever so often.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
@smuzaffar I compiled all CMSSW code that includes `boost/interprocess` and had the headers redirected to this changed copy of boost.

File: boost/interprocess/detail/managed_open_or_create_impl.hpp [hunk 1]
```diff
@@ -310,7 +310,7 @@ class managed_open_or_create_impl
                //File existing when trying to create, but non-existing when
                //trying to open, and tried MaxCreateOrOpenTries times. Something fishy
                //is happening here and we can't solve it
-               throw interprocess_exception(error_info(corrupted_error));
+              throw interprocess_exception(error_info(corrupted_error), "file exists when create but not when open");
```
Comment (fwyzard) [inline boost/interprocess/detail/managed_open_or_create_impl.hpp]:
```suggestion
               throw interprocess_exception(error_info(corrupted_error), "do_create_else_open: file exists when trying to create, but not when trying to open");
```

Comment (fwyzard) [reply boost/interprocess/detail/managed_open_or_create_impl.hpp]: For consistency with the other cases where you mention the function name explicitly.

File: boost/interprocess/managed_shared_memory.hpp [hunk 2]
```diff
@@ -1,4 +1,4 @@
-//////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////
```
Comment (fwyzard) [inline boost/interprocess/managed_shared_memory.hpp]: I guess this could be avoided ?

File: boost/interprocess/sync/posix/semaphore_wrapper.hpp [hunk 3]
```diff
@@ -100,13 +100,13 @@ inline bool semaphore_open
       default:
       {
          error_info err(other_error);
-         throw interprocess_exception(err);
+         throw interprocess_exception(err, "semaphore_open unknown type");
```
Comment (fwyzard) [inline boost/interprocess/sync/posix/semaphore_wrapper.hpp]:
```suggestion
         throw interprocess_exception(err, "semaphore_open: unknown type");
```

File: boost/interprocess/sync/posix/semaphore_wrapper.hpp [hunk 4]
```diff
@@ -100,13 +100,13 @@ inline bool semaphore_open
...
       }
    }
 
    //Check for error
    if(handle == BOOST_INTERPROCESS_POSIX_SEM_FAILED){
-      throw interprocess_exception(error_info(errno));
+     throw interprocess_exception(error_info(errno),"semaphore_open sem_open failed");
```
Comment (fwyzard) [inline boost/interprocess/sync/posix/semaphore_wrapper.hpp]:
```suggestion
      throw interprocess_exception(error_info(errno),"semaphore_open: sem_open failed");
```

File: boost/interprocess/sync/posix/semaphore_wrapper.hpp [hunk 5]
```diff
@@ -148,7 +148,7 @@ inline void semaphore_init(sem_t *handle, unsigned int initialCount)
    //In the future, a successful call might be required to return 0.
    if(ret == -1){
       error_info err = system_error_code();
-      throw interprocess_exception(err);
+      throw interprocess_exception(err, "semaphore_unlink sem_init failed");
```
Comment (fwyzard) [inline boost/interprocess/sync/posix/semaphore_wrapper.hpp]:
```suggestion
      throw interprocess_exception(err, "semaphore_init: sem_init failed");
```

File: boost/interprocess/detail/managed_open_or_create_impl.hpp (hunk 1 above)
Comment (Dr15Jones) [reply boost/interprocess/detail/managed_open_or_create_impl.hpp]: done

File: boost/interprocess/managed_shared_memory.hpp (hunk 2 above)
Comment (Dr15Jones) [reply boost/interprocess/managed_shared_memory.hpp]: done

File: boost/interprocess/sync/posix/semaphore_wrapper.hpp (hunk 3 above)
Comment (Dr15Jones) [reply boost/interprocess/sync/posix/semaphore_wrapper.hpp]: done

File: boost/interprocess/sync/posix/semaphore_wrapper.hpp (hunk 4 above)
Comment (Dr15Jones) [reply boost/interprocess/sync/posix/semaphore_wrapper.hpp]: done

File: boost/interprocess/sync/posix/semaphore_wrapper.hpp (hunk 5 above)
Comment (Dr15Jones) [reply boost/interprocess/sync/posix/semaphore_wrapper.hpp]: done

**Comment by Dr15Jones** [comment]
One thing of note is from looking at the code, it looks like the message `boost::interprocess_exception::library_error` was never supposed to be issued and instead the `error_info` was supposed to give something useful. So this change is just trying to figure out where `error_info` is not working like it was intended and then to use the extra info to figure out our problem.

**Comment by smuzaffar** [comment]
test parameters:
- full_cmssw = true

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
+externals

**Pull request merged**